app/src/main/kotlin/org/eclipse/kuksa/companion/extension/ConstraintScopeExtension.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.eclipse.kuksa.companion.extension
+
+import androidx.compose.material3.windowsizeclass.WindowSizeClass
+import androidx.compose.material3.windowsizeclass.WindowWidthSizeClass
+import androidx.constraintlayout.compose.ConstrainScope
+import androidx.constraintlayout.compose.ConstrainedLayoutReference
+import org.eclipse.kuksa.companion.feature.door.view.horizontalMarginAnchorToDoor
+import org.eclipse.kuksa.companion.feature.door.view.verticalMarginAnchorToBackDoor
+import org.eclipse.kuksa.companion.feature.door.view.verticalMarginAnchorToDoor
+
+fun ConstrainScope.alignDriverFrontDoor(
```

Comment:
This class was the last one I created before putting the ticket to review. Reason should be clear. Remove redundancy and make the "put some element at a specific place (e.g. next to the DriverFrontDoor) without having to copy paste code" as easy as possible. 
 
I probably don't understand the overall picture of your snippet yet, but it looks like it re-introduces redundancy again because each Overlay class needs to manually add their margins / anchorPoints again. I also don't understand the role of the DoorLockOverlayWindow yet. 

Please elaborate or give a more in-depth sample because I agree that - while the chosen solution can easily be changed - it's not optimal because it's much to specific and just clutters the ConstrainScope.

Reply:
I imagined something like this:

It is just refactoring an extension into a class. My main gripe with this extension is that it feels like we should link the RamsesElements with the constraints and not the other way around. It should also be obvious to the devs and easily expandable if we want to add more anchor points of the car (Where extensions feel more hidden here than classes). 

Currently the extension is also available for every compose view with a constraint (which may have nothing to do with Ramses). To have a function like "alignDriverFrontDoor" feels off. 

Maybe you have some ideas to improve this idea even further if you agree. :)

```
// Compose class
val ramsesAnchor = DriverFrontDoorRamsesAnchor(windowSizeClass, anchorPoint)
ramsesAnchor.align(this)

abstract class RamsesAnchor(
    protected val windowSizeClass: WindowSizeClass,
    protected val anchorPoint: ConstrainedLayoutReference,
) {
    fun align(constrainScope: ConstrainScope) {
        when (windowSizeClass.widthSizeClass) {
            WindowWidthSizeClass.Compact -> alignCompact(constrainScope)
            else -> alignMedium(constrainScope)
        }
    }

    protected abstract fun alignCompact(constrainScope: ConstrainScope)

    protected abstract fun alignMedium(constrainScope: ConstrainScope)
}

class DriverFrontDoorRamsesAnchor(windowSizeClass: WindowSizeClass, anchorPoint: ConstrainedLayoutReference) :
    OverlayWindow(windowSizeClass, anchorPoint) {

    override fun alignCompact(constrainScope: ConstrainScope) {
        constrainScope.apply {
            end.linkTo(anchorPoint.start, horizontalMarginAnchorToDoor)
            top.linkTo(anchorPoint.bottom, verticalMarginAnchorToDoor)
        }
    }

    override fun alignMedium(constrainScope: ConstrainScope) {
        constrainScope.apply {
            end.linkTo(anchorPoint.start, verticalMarginAnchorToDoor)
            bottom.linkTo(anchorPoint.top, horizontalMarginAnchorToDoor)
        }
    }
}
```

Reply:
I like this idea (and I implemented it as suggested). Only thing I don't like is the fact that we are using object classes for this eventhough the instances are fixed. 

```
enum class RamsesAnchor {
   DRIVER_FRONT_DOOR, // implement alignCompact and alignMedium
   PASSENGER_FRONT_DOOR,
   DRIVER_BACK_DOOR
   PASSENGER_BACK_DOOR,
   TRUNK,

   fun align(windowSizeClass: WindowSizeClass, anchorPoint: ConstrainedLayoutReference, constrainScope: ConstrainScope) {
...
... rest as before
...
    }
}
```

However I think the signature which you showed feels more natural and sane
Only way to fix this would be by using enums and putting all parameters into the "align" method.